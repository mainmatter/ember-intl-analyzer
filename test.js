const fs = require('fs');
const { run, generateFileList } = require('./index');
const { translationKey } = require('./translation-key');

let fixtures = fs.readdirSync(`${__dirname}/fixtures/`);

describe('Test Fixtures', () => {
  let output;
  let writtenFiles;
  let fixturesWithErrors = [
    'emblem',
    'missing-translations',
    'unused-translations',
    'in-repo-translations',
    'external-addon-translations',
    'concat-expression',
    'chosen-translations',
    'unused-whitelist',
  ];
  let fixturesWithFix = ['remove-unused-translations', 'remove-unused-translations-nested'];
  let fixturesWithConcat = ['concat-expression'];
  let fixturesWithConfig = {
    'external-addon-translations': {
      externalPaths: ['@*/*', 'external-addon'],
    },
    'chosen-translations': {
      translationFiles: ['**/en.yaml'],
    },
    'plugin-config': {
      babelParserPlugins: ['typescript'],
      extensions: ['.ts'],
    },
    'first-class-component-templates': {
      babelParserPlugins: ['typescript'],
      extensions: ['.gts'],
    },
    'unused-whitelist': {
      errorOnUnusedWhitelistEntries: true,
      whitelist: [
        /some\.whitelisted\.translation-(a|b)/,
        /some\.unused\.whitelisted\.translation-(a|b)/,
      ],
    },
    'custom-t-helpers': {
      helpers: ['t-error'],
    },
  };

  beforeEach(() => {
    output = '';
    writtenFiles = new Map();
  });

  function log(str = '') {
    output += `${str}\n`;
  }
  function writeToFile(filePath, updatedTranslations) {
    let fileName = filePath.split('/').slice(-2).join('/');
    writtenFiles.set(fileName, updatedTranslations);
  }

  for (let fixture of fixtures) {
    test(fixture, async () => {
      let returnValue = await run(`${__dirname}/fixtures/${fixture}`, {
        log,
        fix: fixturesWithFix.includes(fixture),
        color: false,
        writeToFile,
        config: fixturesWithConfig[fixture],
        analyzeConcatExpression: fixturesWithConcat.includes(fixture),
      });

      let expectedReturnValue = fixturesWithErrors.includes(fixture) ? 1 : 0;

      expect(returnValue).toEqual(expectedReturnValue);
      expect(output).toMatchSnapshot();
      expect(writtenFiles).toMatchSnapshot();
    });
  }
});

describe('generateFileList', () => {
  const TESTS = [
    [['a.js'], 'a.js'],
    [['a.js', 'b.js'], 'a.js and b.js'],
    [['a.js', 'b.js', 'c.js'], 'a.js, b.js and c.js'],
    [['a.js', 'b.js', 'd.js', 'c.js'], 'a.js, b.js, c.js and d.js'],
    [['translations/en.json', 'translations/de.json'], 'de.json and en.json'],
  ];

  for (let [input, expected] of TESTS) {
    test(expected, () => {
      expect(generateFileList(input)).toBe(expected);
    });
  }

  test('passing an empty array throws an error', () => {
    expect(() => generateFileList([])).toThrow('Unexpected empty file list');
  });
});

describe('translationKey', () => {
  test('util returns the exact same string', () => {
    expect(translationKey`some.translation.key`).toBe('some.translation.key');
  });
});
