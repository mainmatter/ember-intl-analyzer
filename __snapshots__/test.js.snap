// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Fixtures chosen-translations 1`] = `
"[1/4] 🔍  Finding JS and HBS files...
[2/4] 🔍  Searching for translations keys in JS and HBS files...
[3/4] ⚙️   Checking for unused translations...
[4/4] ⚙️   Checking for missing translations...

 👏  No unused translations were found!

 ⚠️   Found 1 missing translations!

   - only-nl-translation (used in app/controllers/application.js)
"
`;

exports[`Test Fixtures chosen-translations 2`] = `Map {}`;

exports[`Test Fixtures concat-expression 1`] = `
"[1/4] 🔍  Finding JS and HBS files...
[2/4] 🔍  Searching for translations keys in JS and HBS files...
[3/4] ⚙️   Checking for unused translations...
[4/4] ⚙️   Checking for missing translations...

 👏  No unused translations were found!

 ⚠️   Found 2 missing translations!

   - prefix.simple-but-missing (used in app/templates/application.hbs)
   - prefix.key-that-should-exist-but-is-missing.value (used in app/templates/application.hbs)
"
`;

exports[`Test Fixtures concat-expression 2`] = `Map {}`;

exports[`Test Fixtures custom-t-helpers 1`] = `
"[1/4] 🔍  Finding JS and HBS files...
[2/4] 🔍  Searching for translations keys in JS and HBS files...
[3/4] ⚙️   Checking for unused translations...
[4/4] ⚙️   Checking for missing translations...

 👏  No unused translations were found!

 👏  No missing translations were found!
"
`;

exports[`Test Fixtures custom-t-helpers 2`] = `Map {}`;

exports[`Test Fixtures decorators 1`] = `
"[1/4] 🔍  Finding JS and HBS files...
[2/4] 🔍  Searching for translations keys in JS and HBS files...
[3/4] ⚙️   Checking for unused translations...
[4/4] ⚙️   Checking for missing translations...

 👏  No unused translations were found!

 👏  No missing translations were found!
"
`;

exports[`Test Fixtures decorators 2`] = `Map {}`;

exports[`Test Fixtures embedded-conditionals 1`] = `
"[1/4] 🔍  Finding JS and HBS files...
[2/4] 🔍  Searching for translations keys in JS and HBS files...
[3/4] ⚙️   Checking for unused translations...
[4/4] ⚙️   Checking for missing translations...

 👏  No unused translations were found!

 👏  No missing translations were found!
"
`;

exports[`Test Fixtures embedded-conditionals 2`] = `Map {}`;

exports[`Test Fixtures emblem 1`] = `
"[1/4] 🔍  Finding JS and HBS files...
[2/4] 🔍  Searching for translations keys in JS and HBS files...
[3/4] ⚙️   Checking for unused translations...
[4/4] ⚙️   Checking for missing translations...

 👏  No unused translations were found!

 ⚠️   Found 1 missing translations!

   - emblem-translation (used in app/templates/application.emblem)
"
`;

exports[`Test Fixtures emblem 2`] = `Map {}`;

exports[`Test Fixtures external-addon-translations 1`] = `
"[1/4] 🔍  Finding JS and HBS files...
[2/4] 🔍  Searching for translations keys in JS and HBS files...
[3/4] ⚙️   Checking for unused translations...
[4/4] ⚙️   Checking for missing translations...

 👏  No unused translations were found!

 ⚠️   Found 1 missing translations!

   - other-external-addon.used-by-app-translation (used in app/templates/application.hbs)
"
`;

exports[`Test Fixtures external-addon-translations 2`] = `Map {}`;

exports[`Test Fixtures first-class-component-templates 1`] = `
"[1/4] 🔍  Finding JS and HBS files...
[2/4] 🔍  Searching for translations keys in JS and HBS files...
[3/4] ⚙️   Checking for unused translations...
[4/4] ⚙️   Checking for missing translations...

 👏  No unused translations were found!

 👏  No missing translations were found!
"
`;

exports[`Test Fixtures first-class-component-templates 2`] = `Map {}`;

exports[`Test Fixtures in-repo-translations 1`] = `
"[1/4] 🔍  Finding JS and HBS files...
[2/4] 🔍  Searching for translations keys in JS and HBS files...
[3/4] ⚙️   Checking for unused translations...
[4/4] ⚙️   Checking for missing translations...

 ⚠️   Found 1 unused translations!
     You can use --fix to remove all unused translations.

   - in-repo-addon.unused (used in lib/in-repo-addon/translations/en.yaml)

 ⚠️   Found 2 missing translations!

   - translation-missing-in-addon (used in lib/in-repo-addon/addon/templates/addon.hbs)
   - translation-missing-in-app (used in lib/in-repo-addon/app/templates/application.hbs)
"
`;

exports[`Test Fixtures in-repo-translations 2`] = `Map {}`;

exports[`Test Fixtures just-t-func 1`] = `
"[1/4] 🔍  Finding JS and HBS files...
[2/4] 🔍  Searching for translations keys in JS and HBS files...
[3/4] ⚙️   Checking for unused translations...
[4/4] ⚙️   Checking for missing translations...

 👏  No unused translations were found!

 👏  No missing translations were found!
"
`;

exports[`Test Fixtures just-t-func 2`] = `Map {}`;

exports[`Test Fixtures missing-translations 1`] = `
"[1/4] 🔍  Finding JS and HBS files...
[2/4] 🔍  Searching for translations keys in JS and HBS files...
[3/4] ⚙️   Checking for unused translations...
[4/4] ⚙️   Checking for missing translations...

 👏  No unused translations were found!

 ⚠️   Found 2 missing translations!

   - js-translation (used in app/controllers/application.js)
   - hbs-translation (used in app/templates/application.hbs)
"
`;

exports[`Test Fixtures missing-translations 2`] = `Map {}`;

exports[`Test Fixtures no-issues 1`] = `
"[1/4] 🔍  Finding JS and HBS files...
[2/4] 🔍  Searching for translations keys in JS and HBS files...
[3/4] ⚙️   Checking for unused translations...
[4/4] ⚙️   Checking for missing translations...

 👏  No unused translations were found!

 👏  No missing translations were found!
"
`;

exports[`Test Fixtures no-issues 2`] = `Map {}`;

exports[`Test Fixtures plugin-config 1`] = `
"[1/4] 🔍  Finding JS and HBS files...
[2/4] 🔍  Searching for translations keys in JS and HBS files...
[3/4] ⚙️   Checking for unused translations...
[4/4] ⚙️   Checking for missing translations...

 👏  No unused translations were found!

 👏  No missing translations were found!
"
`;

exports[`Test Fixtures plugin-config 2`] = `Map {}`;

exports[`Test Fixtures remove-unused-translations 1`] = `
"[1/4] 🔍  Finding JS and HBS files...
[2/4] 🔍  Searching for translations keys in JS and HBS files...
[3/4] ⚙️   Checking for unused translations...
[4/4] ⚙️   Checking for missing translations...

 ⚠️   Found 3 unused translations!

   - foo (used in de.json and en.json)
   - extra (used in de.json)
   - foo.bar.unnested (used in de.json and en.json)

 👏  No missing translations were found!

 👏 All unused translations were removed
"
`;

exports[`Test Fixtures remove-unused-translations 2`] = `
Map {
  "translations/de.json" => "{
  \\"hbs-translation\\": \\"Lenkstage\\",
  \\"js-translation\\": \\"Affentheater\\"
}",
  "translations/en.json" => "{
  \\"hbs-translation\\": \\"HBS!\\",
  \\"js-translation\\": \\"JS!\\"
}",
}
`;

exports[`Test Fixtures remove-unused-translations-nested 1`] = `
"[1/4] 🔍  Finding JS and HBS files...
[2/4] 🔍  Searching for translations keys in JS and HBS files...
[3/4] ⚙️   Checking for unused translations...
[4/4] ⚙️   Checking for missing translations...

 ⚠️   Found 4 unused translations!

   - foo (used in de.json and en.json)
   - extra (used in de.json)
   - parent.child.unusedKey (used in de.json and en.json)
   - unusedParent.unusedChild.unusedKey (used in de.json and en.json)

 👏  No missing translations were found!

 👏 All unused translations were removed
"
`;

exports[`Test Fixtures remove-unused-translations-nested 2`] = `
Map {
  "translations/de.json" => "{
  \\"hbs-translation\\": \\"Lenkstage\\",
  \\"js-translation\\": \\"Affentheater\\",
  \\"parent\\": {
    \\"secondChild\\": \\"usedKey\\"
  }
}",
  "translations/en.json" => "{
  \\"hbs-translation\\": \\"HBS!\\",
  \\"js-translation\\": \\"JS!\\",
  \\"parent\\": {
    \\"secondChild\\": \\"usedKey\\"
  }
}",
}
`;

exports[`Test Fixtures standalone-translation-key 1`] = `
"[1/4] 🔍  Finding JS and HBS files...
[2/4] 🔍  Searching for translations keys in JS and HBS files...
[3/4] ⚙️   Checking for unused translations...
[4/4] ⚙️   Checking for missing translations...

 👏  No unused translations were found!

 👏  No missing translations were found!
"
`;

exports[`Test Fixtures standalone-translation-key 2`] = `Map {}`;

exports[`Test Fixtures unused-translations 1`] = `
"[1/4] 🔍  Finding JS and HBS files...
[2/4] 🔍  Searching for translations keys in JS and HBS files...
[3/4] ⚙️   Checking for unused translations...
[4/4] ⚙️   Checking for missing translations...

 ⚠️   Found 2 unused translations!
     You can use --fix to remove all unused translations.

   - foo (used in de.json and en.json)
   - extra (used in de.json)

 👏  No missing translations were found!
"
`;

exports[`Test Fixtures unused-translations 2`] = `Map {}`;

exports[`Test Fixtures unused-whitelist 1`] = `
"[1/4] 🔍  Finding JS and HBS files...
[2/4] 🔍  Searching for translations keys in JS and HBS files...
[3/4] ⚙️   Checking for unused translations...
[4/4] ⚙️   Checking for missing translations...

 👏  No unused translations were found!

 👏  No missing translations were found!

 ⚠️   Found 1 unused whitelist entry! Please remove it from the whitelist:

   - /some\\\\.unused\\\\.whitelisted\\\\.translation-(a|b)/
"
`;

exports[`Test Fixtures unused-whitelist 2`] = `Map {}`;

exports[`Test Fixtures yaml-translations 1`] = `
"[1/4] 🔍  Finding JS and HBS files...
[2/4] 🔍  Searching for translations keys in JS and HBS files...
[3/4] ⚙️   Checking for unused translations...
[4/4] ⚙️   Checking for missing translations...

 👏  No unused translations were found!

 👏  No missing translations were found!
"
`;

exports[`Test Fixtures yaml-translations 2`] = `Map {}`;
